Introduce compile-time checking and alternate allocation for compact pointer types

ddegazio opened the pull request:
#### 27959d8049add866e8696984338a697f5419449d
<pre>
Introduce compile-time checking and alternate allocation for compact pointer types
<a href="https://bugs.webkit.org/show_bug.cgi?id=266080">https://bugs.webkit.org/show_bug.cgi?id=266080</a>
<a href="https://rdar.apple.com/119382681">rdar://119382681</a>

Reviewed by Mark Lam.

Adds compile-time checking to WTF compact and packed pointer types requiring the
pointed-to types to be explicitly marked as compatible with compact pointers. Also
adds new equivalents to the existing WTF::fastMalloc and WTF::isoMalloc allocation
functions for compactible types, giving us an API point to play with how types
used in compact pointers are allocated.

* Source/JavaScriptCore/bytecode/CallLinkInfoBase.h:
* Source/JavaScriptCore/bytecode/CodeBlock.h:
* Source/JavaScriptCore/bytecode/InlineCallFrame.h:
* Source/JavaScriptCore/bytecode/InlineCallFrameSet.h:
* Source/JavaScriptCore/bytecode/StructureStubClearingWatchpoint.h:
* Source/JavaScriptCore/bytecode/Watchpoint.h:
* Source/JavaScriptCore/dfg/DFGJITCode.h:
* Source/JavaScriptCore/heap/FastMallocAlignedMemoryAllocator.cpp:
* Source/JavaScriptCore/heap/IsoAlignedMemoryAllocator.cpp:
* Source/JavaScriptCore/jit/BaselineJITCode.h:
* Source/JavaScriptCore/jit/ExecutableMemoryHandle.h:
* Source/JavaScriptCore/jit/JITCode.h:
* Source/JavaScriptCore/jit/PolymorphicCallStubRoutine.h:
* Source/JavaScriptCore/runtime/ClassInfo.h:
* Source/JavaScriptCore/runtime/JSArray.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
* Source/WTF/wtf/Bag.h:
* Source/WTF/wtf/CompactPointerTuple.h:
* Source/WTF/wtf/CompactPtr.h:
* Source/WTF/wtf/DebugHeap.h:
* Source/WTF/wtf/FastMalloc.cpp:
* Source/WTF/wtf/FastMalloc.h:
* Source/WTF/wtf/IsoMalloc.h:
* Source/WTF/wtf/IsoMallocInlines.h:
* Source/WTF/wtf/SentinelLinkedList.h:
* Source/WTF/wtf/WeakPtr.h:
* Source/WTF/wtf/text/StringImpl.h:
* Source/WTF/wtf/text/WTFString.h:
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/NodeRareData.h:
* Source/WebCore/rendering/RenderObject.cpp:
* Source/WebCore/rendering/RenderObject.h:
* Source/bmalloc/bmalloc/AllocationCounts.h:
* Source/bmalloc/bmalloc/IsoHeap.cpp:
* Source/bmalloc/bmalloc/IsoHeap.h:
* Source/bmalloc/bmalloc/IsoHeapImpl.h:
* Source/bmalloc/bmalloc/IsoHeapInlines.h:
* Source/bmalloc/bmalloc/Packed.h:
* Tools/TestWebKitAPI/Tests/WTF/AlignedRefLogger.h:
* Tools/TestWebKitAPI/Tests/WTF/CompactRefPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/CompactUniquePtrTuple.cpp:
* Tools/TestWebKitAPI/Tests/WTF/RefLogger.h:

Canonical link: <a href="https://commits.webkit.org/271796@main">https://commits.webkit.org/271796@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/24477dc8e8961fff9e698399d067478ef5328644

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29703 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8356 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30998 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32198 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26857 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30319 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10504 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5630 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29975 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6969 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25326 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5948 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/6089 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26414 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33538 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/25491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27065 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26838 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/32296 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/29900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/6035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4233 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/30073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7779 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/36267 "Built successfully") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/6578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7823 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3822 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6568 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/heap/FastMallocAlignedMemoryAllocator.cpp:
```diff
@@ -46,7 +46,7 @@ void* FastMallocAlignedMemoryAllocator::tryAllocateAlignedMemory(size_t alignmen
     return m_heap.memalign(alignment, size, true);
 #else
-    return tryFastAlignedMalloc(alignment, size);
+    return tryFastAlignedMallocCompact(alignment, size);
```
Let's call this `tryFastCompactAlignedMalloc` to be consistent with the rest.

MenloDorian commented on Source/JavaScriptCore/heap/IsoAlignedMemoryAllocator.cpp:
```diff
@@ -76,7 +76,7 @@ void* IsoAlignedMemoryAllocator::tryMallocBlock()
     return m_heap.memalign(MarkedBlock::blockSize, MarkedBlock::blockSize, true);
 #else
-    return tryFastAlignedMalloc(MarkedBlock::blockSize, MarkedBlock::blockSize);
+    return tryFastAlignedMallocCompact(MarkedBlock::blockSize, MarkedBlock::blockSize);
```
Ditto: `tryFastCompactAlignedMalloc`

MenloDorian commented on Source/JavaScriptCore/jit/JITCode.h:
```diff
@@ -72,6 +72,41 @@ enum class JITType : uint8_t {
 static constexpr unsigned widthOfJITType = 3;
 static_assert(WTF::getMSBSetConstexpr(static_cast<std::underlying_type_t<JITType>>(JITType::FTLJIT)) + 1 == widthOfJITType);
 
+template<typename ConstantType>
```
Let's rename `ConstantType` to `ByteSizedEnumType` or something like that.  That makes it clear why sizeof must be 1, and avoids confusion.

MenloDorian commented on Source/bmalloc/bmalloc/IsoHeap.h:
```diff
@@ -49,14 +49,43 @@ namespace api {
+
+    void scavenge()
+    {
+    }
```
nit: You can also express this more compactly as:
```
void scavenge() { }
```

It enables more denser info at a glance especially since these are all empty or single statement functions.  Note: this single line style is not permitted for functions with multiple statements.

MenloDorian reviewed (approved):
r=me with fixes.  Please also resolve EWS failure issues, and also coordinate with Michael Saboff on his TZone patch.  I suspect your patch will have to be adapted a bit if his lands first.

ddegazio committed:
Introduce compile-time checking and alternate allocation for compact pointer types
https://bugs.webkit.org/show_bug.cgi?id=266080
rdar://119382681

Reviewed by Mark Lam.

Adds compile-time checking to WTF compact and packed pointer types requiring the
pointed-to types to be explicitly marked as compatible with compact pointers. Also
adds new equivalents to the existing WTF::fastMalloc and WTF::isoMalloc allocation
functions for compactible types, giving us an API point to play with how types
used in compact pointers are allocated.

* Source/JavaScriptCore/bytecode/CallLinkInfoBase.h:
* Source/JavaScriptCore/bytecode/CodeBlock.h:
* Source/JavaScriptCore/bytecode/InlineCallFrame.h:
* Source/JavaScriptCore/bytecode/InlineCallFrameSet.h:
* Source/JavaScriptCore/bytecode/StructureStubClearingWatchpoint.h:
* Source/JavaScriptCore/bytecode/Watchpoint.h:
* Source/JavaScriptCore/dfg/DFGJITCode.h:
* Source/JavaScriptCore/heap/FastMallocAlignedMemoryAllocator.cpp:
* Source/JavaScriptCore/heap/IsoAlignedMemoryAllocator.cpp:
* Source/JavaScriptCore/jit/BaselineJITCode.h:
* Source/JavaScriptCore/jit/ExecutableMemoryHandle.h:
* Source/JavaScriptCore/jit/JITCode.h:
* Source/JavaScriptCore/jit/PolymorphicCallStubRoutine.h:
* Source/JavaScriptCore/runtime/ClassInfo.h:
* Source/JavaScriptCore/runtime/JSArray.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
* Source/WTF/wtf/Bag.h:
* Source/WTF/wtf/CompactPointerTuple.h:
* Source/WTF/wtf/CompactPtr.h:
* Source/WTF/wtf/DebugHeap.h:
* Source/WTF/wtf/FastMalloc.cpp:
* Source/WTF/wtf/FastMalloc.h:
* Source/WTF/wtf/IsoMalloc.h:
* Source/WTF/wtf/IsoMallocInlines.h:
* Source/WTF/wtf/SentinelLinkedList.h:
* Source/WTF/wtf/WeakPtr.h:
* Source/WTF/wtf/text/StringImpl.h:
* Source/WTF/wtf/text/WTFString.h:
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/NodeRareData.h:
* Source/WebCore/rendering/RenderObject.cpp:
* Source/WebCore/rendering/RenderObject.h:
* Source/bmalloc/bmalloc/AllocationCounts.h:
* Source/bmalloc/bmalloc/IsoHeap.cpp:
* Source/bmalloc/bmalloc/IsoHeap.h:
* Source/bmalloc/bmalloc/IsoHeapImpl.h:
* Source/bmalloc/bmalloc/IsoHeapInlines.h:
* Source/bmalloc/bmalloc/Packed.h:
* Tools/TestWebKitAPI/Tests/WTF/AlignedRefLogger.h:
* Tools/TestWebKitAPI/Tests/WTF/CompactRefPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/CompactUniquePtrTuple.cpp:
* Tools/TestWebKitAPI/Tests/WTF/RefLogger.h:

Canonical link: https://commits.webkit.org/271796@main
